presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayoutForPredicate.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.optimizations.TableLayoutRewriter;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.tree.BooleanLiteral;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Pick a table layout based on the filter above the TableScanNode,
+ * if a table layout has not yet been assigned.
+ */
+public class PickTableLayoutForPredicate
+        implements Rule
+{
+    private static final Pattern PATTERN = Pattern.typeOf(FilterNode.class);
+
+    private final Metadata metadata;
+
+    public PickTableLayoutForPredicate(Metadata metadata)
+    {
+        requireNonNull(metadata, "metadata is null");
+
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Context context)
+    {
+        if (!(node instanceof FilterNode)) {
+            return Optional.empty();
+        }
+
+        FilterNode filterNode = (FilterNode) node;
+        PlanNode source = context.getLookup().resolve(filterNode.getSource());
+
+        if (!((source instanceof TableScanNode) && shouldRewriteTableLayout((TableScanNode) source))) {
+            return Optional.empty();
+        }
+
+        TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator());
+        PlanNode rewrittenTableScan = tableLayoutRewriter.planTableScan((TableScanNode) source, filterNode.getPredicate());
+        if (!node.equals(rewrittenTableScan)) {
```

Comment:
I do not think you can use `equals`, I would not expect it is implemented correctly.

Other than that, I think it should be a responsibility of `IterativeOptimizer` to treat such case when apply returns the same node as it got in input in the same way as it would `Optional.empty()`

Also I checked the rewriter and it does not look as it could return the same node.